miasm2/analysis/mem.py

```diff
@@ -0,0 +1,1259 @@
+        cls._attrs = {}
+        offset = 0
+        for name, field in cls.fields:
+            # For reflexion
+            field._set_self_type(cls)
+            cls.gen_field(name, field, offset)
+            offset += field.size()
+        cls._size = offset
+
+    @classmethod
+    def gen_field(cls, name, field, offset):
+        """Generate only one field
+
+        Args:
+            @name: (str) the name of the field
+            @field: (MemField instance) the field type
+            @offset: (int) the offset of the field in the structure
+        """
+        cls._gen_simple_attr(name, field, offset)
+        if isinstance(field, Union):
+            cls._gen_union_attr(field, offset)
+
+    @classmethod
+    def _gen_simple_attr(cls, name, field, offset):
+        cls._attrs[name] = {"field": field, "offset": offset}
+
+        # Generate self.<name> getter and setter
+        setattr(cls, name, property(
+            lambda self: self.get_field(name),
+            lambda self, val: self.set_field(name, val)
+        ))
+
+        # Generate self.deref_<name> getter and setter if this field is a
+        # Ptr
+        if isinstance(field, Ptr):
+            setattr(cls, "deref_%s" % name, property(
+                lambda self: self.deref_field(name),
+                lambda self, val: self.set_deref_field(name, val)
+            ))
+
+    @classmethod
+    def _gen_union_attr(cls, union_field, offset):
+        if not isinstance(union_field, Union):
+            raise ValueError("field should be an Union instance")
+        for name, field in union_field.field_list:
+            cls.gen_field(name, field, offset)
+
+
+class MemSelf(MemStruct):
+    """Special Marker class for reference to current class in a Ptr or Array
+    (mostly Array of Ptr).
+
+    Example:
+        class ListNode(MemStruct):
+            fields = [
+                ("next", Ptr("<I", MemSelf)),
+                ("data", Ptr("<I", MemVoid)),
+            ]
+    """
+    pass
+
+
+class MemVoid(MemStruct):
+    """Placeholder for e.g. Ptr to an undetermined type. Useful mostly when
+    casted to another type. Allows to implement C's "void*" pattern.
+    """
+    def __repr__(self):
+        return self.__class__.__name__
+
+
+# This does not use _MetaMemStruct features, impl is custom for strings,
+# because they are unsized. The only memory field is self.value.
+class MemStr(MemStruct):
+    """Implements a string representation in memory.
+
+    The @encoding is passed to the constructor, and is currently either null
+    terminated "ansi" (latin1) or (double) null terminated "utf16". Be aware
+    that the utf16 implementation is a bit buggy...
+
+    The string value can be got or set (with python str/unicode) through the
+    self.value attribute. String encoding/decoding is handled by the class.
+
+    This type is dynamically sized only (get_size is implemented, not sizeof).
+    """
+    def __init__(self, vm, addr, encoding="ansi"):
```

Comment:
Or, maybe, you can use subclasses for each encoding and a registration in the main class (`MemStr`).
This is the pattern used in `Container` and `Translators`, which seems to be cleaner than a list of conditions in each method. In other words, you are doing the split inside methods, but it can be done by the class.